File: 2020/python/day6.py
#!/usr/bin/env python3
import sys


def day6(input_file):
    with open(input_file, "r") as f:
        input = f.read()

    print("Day 6: Custom Customs")
    print(f"\tPart 1: {__part1(input)}")
    print(f"\tPart 2: {__part2(input)}")


def __part1(input):
    sum = 0

    groups = input.split("\n\n")
    for group in groups:
        map = {}
        for line in group.splitlines():
            for c in line:
                map[c] = True
        sum += len(map.keys())

    return sum


def __part2(input):
    sum = 0

    groups = input.split("\n\n")
    for group in groups:
        map = {}
        for line in group.splitlines():
            for c in line:
                try:
                    map[c] += 1
                except KeyError:
                    map[c] = 1
        all = 0
        groupsize = len(group.splitlines())
        for k, v in map.items():
            if v == groupsize:
                all += 1

        sum += all

    return sum


if __name__ == "__main__":
    day6(sys.argv[1])

File: 2020/javascript/day6.js
#!/usr/bin/env node

const fs = require("fs");

const part1 = (input) => {
  return input.split("\n\n").reduce((sum, group) => {
    return (
      sum +
      Object.keys(
        group.split("\n").reduce((set, line) => {
          return {
            ...[...line].reduce((qs, c) => {
              return { ...qs, [c]: true };
            }, set),
          };
        }, {})
      ).length
    );
  }, 0);
};

const part2 = (input) => {
  return input.split("\n\n").reduce((sum, group) => {
    return (
      sum +
      Object.entries(
        group.split("\n").reduce((set, line) => {
          return {
            ...[...line].reduce((qs, c) => {
              return { ...qs, [c]: qs[c] ? qs[c] + 1 : 1 };
            }, set),
          };
        }, {})
      ).filter(([k, v]) => v == group.split("\n").length).length
    );
  }, 0);
};

const day6 = (inputFile) => {
  const input = fs.readFileSync(inputFile, "utf8");

  console.log("Day 6: Custom Customs");
  console.log(`\tPart 1: ${part1(input)}`);
  console.log(`\tPart 2: ${part2(input)}`);
};

module.exports = day6;

if (require.main === module) day6(process.argv[2]);
